### ISSUE TITLE
Order of triangles to rasterization matters

### ISSUE THREAD

**Issue by Janiels** [recastnavigation/recastnavigation#232]
I am currently having a case where the result look like this:
![walkable voxels before](https://cloud.githubusercontent.com/assets/7887810/18483803/093f8bde-79e6-11e6-9a67-72d71b5cf66f.png)
If I shuffle around the triangles a bit so spans are added in a different order, the result looks like this:
![walkable voxels after](https://cloud.githubusercontent.com/assets/7887810/18483823/1bff50ec-79e6-11e6-96da-5aff6d6ae1f0.png)

The order of the spans added looks like this for the two cases:
```
Mesh over ledge case:
Adding 682-684 of 0
Adding 548-549 of 0
Adding 625-626 of 63
Adding 685-686 of 63
Adding 685-687 of 0
Adding 686-688 of 0
Adding 765-767 of 63
Adding 765-767 of 63
Adding 761-763 of 0
Adding 69-70 of 63
Result:
69-70 of 63
548-549 of 0
625-626 of 63
682-684 of 0
685-688 of 63
761-763 of 0
765-767 of 63
```

```
No mesh over ledge case:
Adding 69-70 of 63
Adding 765-767 of 63
Adding 765-767 of 63
Adding 682-684 of 0
Adding 761-763 of 0
Adding 548-549 of 0
Adding 625-626 of 63
Adding 685-686 of 63
Adding 686-688 of 0
Adding 685-687 of 0
Result:
69-70 of 63
548-549 of 0
625-626 of 63
682-684 of 0
685-688 of 0
761-763 of 0
765-767 of 63
```

The discrepancy is easy to see in the 685-688 span.

**Comment by Janiels** [comment]
The error looks to be around here: https://github.com/recastnavigation/recastnavigation/blob/15ebb8bd25e952215d2ff140d5ec01fa50294275/Recast/Source/RecastRasterization.cpp#L128-L133
The flags should be merged before `smax` has been merged; otherwise the difference will always be 0 if the previous span ended before the new span ends.
@memononen Does that sound correct?

**Comment by Janiels** [comment]
No, that will not fix the issue here. There is a more fundamental issue of doing the merging this way. Let's assume `flagMergeThr = 1`; then if we have intersecting spans with smax
1 2 3
they will end up with the max area of 1, 2, 3, while
1 3 2
will end up with the max area of 1, 2. This is regardless of changing the behavior as I described in previous post.

I'm not quite sure how that could even be fixed though. Perhaps something that "wraps" spans of lower area type around spans of higher area type?

**Comment by memononen** [comment]
If I look at the data correctly, this happens:
```
Mesh over ledge case:
Adding 5-6 of 63
Adding 5-7 of 0
Adding 6-8 of 0
Result:
5-6 of 63
5-7 of 63
5-8 of 63

No mesh over ledge case:
Adding 5-6 of 63
Adding 6-8 of 0
Adding 5-7 of 0
Result:
5-6 of 63
5-8 of 0
5-8 of 0
```
I'm kind of surprised this has not happened more.

I cannot think of any other solution than to only allow to merge spans of same type during `addSpan()` and then do final merge when converting to CHF.

**Comment by Janiels** [comment]
How do you think intersecting segments of different types should be handled?
Should they both be stored (probably ordered by smax only)? Or should we do a tie breaker for the intersecting parts (eg. always prefer highest area type or something like that)?

**Comment by memononen** [comment]
I think it would be a good idea to have the segments as sorted list based on `smin`. And the final merge should be the same as inserting each span in `smin` order using the current algorithm.

Doing merge while rasterizing gets rid of a lot of spans. We might end up having 2x or so spans.

**Comment by Janiels** [comment]
Ok, I will try to play around with that. However I'm having a hard time wrapping my head around the problem we need to solve here.

I realize that there can be multiple triangles with different flags touching the same voxels. That is the fundamental issue, and we need some way to handle this situation, and to properly end up with the correct flags in the final heightfield.
To me it seems a good solution should be to always pick the flag of the top-most span. Since we are merging intersecting spans, there is a solid column of terrain here, and we will only generate mesh on top; so pick the top flag.
Then we just need a rule for which flag to pick when there is no top-most span. Similarly to `flagMergeThr` we could just pick the max flag. It seems pretty arbitrary and the correct solution would probably be case-by-case depending on the input terrain, but picking max is ok.
With this algorithm, we get the following:
```
Mesh over ledge case:
Adding 5-6 of 63
Adding 5-7 of 0
Adding 6-8 of 0
Result:
5-6 of 63
5-7 of 0
5-8 of 0

No mesh over ledge case:
Adding 5-6 of 63
Adding 6-8 of 0
Adding 5-7 of 0
Result:
5-6 of 63
5-8 of 0
5-8 of 0
```
Both cases are equal, and it makes sense with the rasterization input (the triangles here are too steep).

The only problem left is to properly handle the climbing here. To me it seems `rcFilterLowHangingWalkableObstacles` is the function that is supposed to handle this, but it does not check if we can climb from neighbors, only if there is a walkable span below that is within the climb height. So I don't quite understand how that works reliably at all. Do I misunderstand something here, and how does my rasterization change sound?

**Comment by memononen** [comment]
The idea behind picking max area type is to favor walkable areas when there's walkable and non-wealkable area close to each other.

The reason for the current merge logic (take largest) is that many games have small geometry, which is not navigable based on slope. Think lo-res pebble or tree roots. Or even the vertical section of stairs. There were even more exotic cases on production levels. `rcFilterLowHangingWalkableObstacles` handles the same problem in case the obstacle is floating. This happens more when the cell height is small.

An earlier version of Recast tried to reconstruct the slope from the rasterized data, but It was not very reliable. That would have solved the root cause (pun intended).

**Comment by Janiels** [comment]
So where is climbing actually handled? Because `rcFilterLowHangingWalkableObstacles` and the flag merging in rasterization only handles it within the same cell. Eg. if there is no walkable span below something that is obviously climbable it will be marked unnavigable due to slope. I assume this is just an approximation that turns out to work in practice and is efficient?

I checked one of your old blog posts: http://digestingduck.blogspot.dk/2010/01/rough-fringes.html
Ignoring efficiency, what stops us from marking neighbors within climb height of walkable neighbors as walkable? The only thing I could think of is the problem that should already be solved in `rcFilterLedgeSpans` under the Filtering Neighbours section in that post.

**Comment by memononen** [comment]
`rcFilterLowHangingWalkableObstacles` operates on two consecutive spans. If the bottom is walkable and top is not, and the top of spans are within threshold, then the top will be walkable too. Flag merging handles the case when the spans overlap.

http://4.bp.blogspot.com/_-u6ZJlBFOL0/S2K_jaUWsHI/AAAAAAAAAKc/c2Rx7Rdwxgw/s400/fringe_bad_voxels.jpg
Left: span merging, right: rcFilterLowHangingWalkableObstacles

"...marking neighbors within climb height of walkable neighbors as walkable?"
Does not compute :)

**Comment by Janiels** [comment]
![climbing neighbors](https://cloud.githubusercontent.com/assets/7887810/18546753/edf1e966-7b3f-11e6-89bd-5cf9872fe5b1.png)
I don't know if this helps. My Paint skills are not good :smile:.
Essentially, as far as I can tell, the climbing right now relies on there being some walkable terrain below the step. But this is not guaranteed. In my picture there, there is a span touching just one steep triangle. So it is marked unwalkable, even though the agent can climb over it. And since there is no terrain below it, it is never promoted to walkable.

**Comment by memononen** [comment]
Got it :)
I think that kind of filter is doable. It is possible that it will ran into false positives. Digging out neighbour information of heightfield is not fun, but doable.

**Comment by Janiels** [comment]
![continuous promotion](https://cloud.githubusercontent.com/assets/7887810/18547321/1b78ada4-7b43-11e6-89fd-60b627b73700.png)
Here is a possible false positive. But this one should already be handled in `rcFilterLedgeSpans`, correct?

But if we could use this kind of algorithm we could use the simple deterministic rasterization algorithm I mentioned, and get rid of `rcFilterLowHangingWalkableObstacles` too.

**Comment by memononen** [comment]
If the filter is set up so that it will try to "fill in" only if the height diff between left and right is less than climb, then that above should not trigger the fill in.

I don't see how your suggested filter will handle `rcFilterLowHangingWalkableObstacles`. Or solve the "always use top" merging.

Single voxel sized non-walkable areas is one thing, but there are many others. I.e. the diagonal bar in the obstacle A (in the blog post) has two pixel wide gaps. Sometimes you have double sided geometry, i.e. non-walkable and walkable on top of each other, etc.

I still think that the general solution is to only merge spans of same type and then do a final pass to merge the flags (maybe even together with low hanging spans).

**Comment by Janiels** [comment]
> If the filter is set up so that it will try to "fill in" only if the height diff between left and right is less than climb, then that above should not trigger the fill in.
I was concerned that the walkability flag may be migrated from span to span (i.e. bottom is walkable means we can climb one voxel up, next voxel is walkable means we can climb another up, etc.). I am probably not thinking in correct terms of how it would need to be implemented.

> I don't see how your suggested filter will handle rcFilterLowHangingWalkableObstacles.
![low hanging](https://cloud.githubusercontent.com/assets/7887810/18548009/b3fbd576-7b46-11e6-8e18-68a35def89f1.png)

> Or solve the "always use top" merging.
Mostly same way as previous case. If top is unnavigable because of a steep triangle, it will be marked navigable if we can climb onto it.

> Single voxel sized non-walkable areas is one thing, but there are many others. I.e. the diagonal bar in the obstacle A (in the blog post) has two pixel wide gaps. Sometimes you have double sided geometry, i.e. non-walkable and walkable on top of each other, etc.
Yes, it would need to be a more general solution, not just the 1-pixel wide version. I haven't even figured out when we should be able to climb over unnavigable voxels. When it has a walkable neighbor within agent radius that is within climb distance? I am not sure how climbing is usually performed in physics engines...

I haven't worked a lot with this data before so I probably still have some fundamental misunderstandings about it :)

**Comment by hymerman** [comment]
Quick note; in my experience the 'climbing' is really 'stepping' and wouldn't be handled with e.g. bespoke climbing animations, but normal locomotion with IK to place the feet and alter pose. Actual climbing needs something more, e.g. off-mesh links. Physics-wise, on a basic level it's handled by bumping a capsule collider over the physics geometry (with an end cap radius of the climb height), but that's quite unreliable so many games use extra raycasts around characters to move the collider up/down if they need to step.

**Comment by Janiels** [comment]
@hymerman Thanks for the explanation. Do you know how this particular case of stepping is handled (with a non-vertical triangle but still steppable)? Is the check for flat terrain in 2D range of agent radius, below stepping height?

**Comment by hymerman** [comment]
TL;DR: there are many ways, some will treat your example as walkable, some will not :( We can either be conservative and cause a break in the navmesh, demand good physics from users and join it up, or provide options for users to configure it to match the navmesh to how their physics works.

----

The situation in your last diagram with the floating object? Annoyingly, it depends on the physics implementation - some games will happily move a character up that, some won't. As such it's hard to know what to do with the navmesh :(

The *best* implementations (i.e. the ones that seem to map capsules most accurately to what humans would actually be able to do given the geometry) will indeed have a several raycast checks for flat terrain in 2D range (not necessarily agent radius, maybe a bit further), within step height, with a check that there's line of sight to that point so they're not stepping on terrain the other side of a thin wall. As such, the obstacle on the right in your example would be traversable.

The *most common* implementation (i.e. the one everyone starts with and is all over the web in physics character controller tutorials) is applying lateral forces to a capsule with end cap radius ~= agent radius, and allowing contact resolution to move the character in the right direction. In that case, in your example the object probably won't be traversable as the capsule will probably collide with the slope first rather than the corner with the flat part, and that slope is steep enough that gravity will counteract the upwards component of the force.

So here it looks like we need to either

- Be conservative and put a break in the navmesh there, so we don't generate paths that some codebases can't navigate
- Be a bit opinionated about how end users should make characters follow these paths, joining the navmesh in these cases and telling people they should have better character controllers
- Provide options to handle this in different ways

Sorry, I feel like I've just made it more complicated than it was! Just thought I'd offer some observations about how I've seen things done out 'in the wild'. Take it with a pinch of salt, clearly I've only seen a tiny proportion of all games' implementations :)

**Comment by Janiels** [comment]
@hymerman Thanks so much for that. I think the best course of action is to treat it as a separate issue from this PR, and if we want to treat it, it should just be another filter that can be called/not called depending on what you want.

In that case the issues will probably be pretty orthogonal. I will look into doing what Mikko suggested (with a final merging step doing CHF construction). If it turns out to be too slow/memory intensive I will probably not bother at all as it seems sufficiently rare.

Thanks for the inputs all around!

**Comment by hymerman** [comment]
:+1: Good call!

**Comment by Janiels** [comment]
Actually I may have been too fast. The floating obstacle is already handled by `rcFilterLowHangingWalkableObstacles`. The case I am talking about is this one:
![climbing neighbors](https://cloud.githubusercontent.com/assets/7887810/18583293/0d69096c-7c0a-11e6-8164-950e84701fb1.png)
But I guess the same principle applies?

**Comment by hymerman** [comment]
Yep, same thing physics- and animation-wise.

**Comment by grahamboree** [comment]
My 2 cents: I think what Mikko outlined is a step in the right direction here: merge only same-flag spans, and then perform a final merge step when generating the chf.  In that final merge step, though, we still have the same problem: 
e.g.
```
            |________Flag_0_________|
|_______________Flag_1_________________|
```
could result in 
```
|___1___|__________0_____________|__1___|
```
or
```
|__________________1___________________|
```
depending on the order in which spans with different flags are combined.  Is this something we'd decree some logic on in the docs (e.g. prefer higher span flag numbers) and/or should we allow the user to control this?  Seems like it has the potential of having a serious impact on the final result.  Does combining it from highest to lowest smax solve this?

**Comment by Janiels** [comment]
@grahamboree It solves it in the way that it does something consistent, regardless of the order the triangles are passed in. However, it would retain the logic to always prefer the higher area types.  Might be good to document that - but I don't really think there is a good clean way to allow users to pick the tie breakers on a case-by-case basis.

**Comment by grahamboree** [comment]
@Janiels Yeah I think you're right.  The only other thing we really could do is allow the user to specify something like "always prefer areas with id 1 over 5" but I'm not sure how useful that really is.  Keeping things more or less consistent with what we've already got, and documenting this behavior is probably the best bet.

**Comment by Daerst** [comment]
Here's an error that's happening both in some custom code using Recast, as well as in latest Unity, with three simple cubes:
![grafik](https://user-images.githubusercontent.com/2000550/111493381-c7d03d00-873d-11eb-9175-73326f1c93a1.png)

The cube on the bottom has AreaID = Water (think e.g. an ocean), the large 'tower' block and the slope leading up to it have AreaID = Walkable. In our custom code we use 0 = NonWalkable, 1 = Walkable, 2 = Water. Unity uses 0 = Walkable, 1 = NonWalkable, and then 3 is the first user-defined area, though they're probably doing some internal remapping to play well with Recast's internal RC_NULL_AREA = 0.

The issue that I'm seeing is that the Water area from below is being merged / propagated up all the way along the non-walkable upright triangle, where it even takes precedence over the Walkable because it has the larger ID:
```csharp
// Merge flags. 
if (rcAbs((int)s->smax - (int)cur->smax) <= flagMergeThr) 
    s->area = rcMax(s->area, cur->area); 
```

I 'fixed' this by giving Walkable a large ID, but of course that's more of a workaround than a general solution.
